services/shhext/mailservers/connmanager.go

```diff
@@ -0,0 +1,202 @@
+				log.Error("retry after error to subscribe for whisper events", "error", err)
+				whisperSub = ps.whisper.SubscribeEnvelopeEvents(whisperEvents)
+			case newNodes := <-ps.notifications:
+				replace := map[enode.ID]*enode.Node{}
+				for _, n := range newNodes {
+					replace[n.ID()] = n
+				}
+				replaceNodes(ps.server, ps.connectedTarget, connected, current, replace)
+				current = replace
+			case ev := <-events:
+				switch ev.Type {
+				case p2p.PeerEventTypeAdd:
+					log.Debug("connected to a mailserver", "address", ev.Peer)
+					nodeAdded(ps.server, ev.Peer, ps.connectedTarget, connected, current)
+				case p2p.PeerEventTypeDrop:
+					log.Debug("mailserver disconnected", "address", ev.Peer)
+					nodeDisconnected(ps.server, ev.Peer, ps.connectedTarget, connected, current)
+				}
+			case ev := <-whisperEvents:
+				// TODO what about completed but with error? what about expired envelopes?
+				switch ev.Event {
+				case whisperv6.EventMailServerRequestSent:
+					requests[ev.Hash] = struct{}{}
+				case whisperv6.EventMailServerRequestCompleted:
+					delete(requests, ev.Hash)
+				case whisperv6.EventMailServerRequestExpired:
+					_, exist := requests[ev.Hash]
+					if !exist {
+						continue
+					}
+					log.Debug("request to a mail server expired, disconnet a peer", "address", ev.Peer)
+					nodeDisconnected(ps.server, ev.Peer, ps.connectedTarget, connected, current)
+				}
+			}
+		}
+	}()
+}
+
+// Stop gracefully closes all background goroutines and waits until they finish.
+func (ps *ConnectionManager) Stop() {
+	if ps.quit == nil {
+		return
+	}
+	select {
+	case <-ps.quit:
+		return
+	default:
+	}
+	close(ps.quit)
+	ps.wg.Wait()
+	ps.quit = nil
+}
+
+func replaceNodes(srv PeerAdderRemover, target int, connected map[enode.ID]struct{}, old, new map[enode.ID]*enode.Node) {
+	for nid, n := range old {
+		if _, exist := new[nid]; !exist {
+			if _, exist := connected[nid]; exist {
+				delete(connected, nid)
+			}
+			srv.RemovePeer(n)
+		}
+	}
+	if len(connected) < target {
+		for _, n := range new {
+			srv.AddPeer(n)
+		}
+	}
+}
+
+func nodeAdded(srv PeerAdderRemover, peer enode.ID, target int, connected map[enode.ID]struct{}, nodes map[enode.ID]*enode.Node) {
+	n, exist := nodes[peer]
+	if !exist {
+		return
+	}
+	if len(connected) == target {
+		srv.RemovePeer(n)
+	} else {
+		connected[n.ID()] = struct{}{}
+	}
+}
+
+func nodeDisconnected(srv PeerAdderRemover, peer enode.ID, target int, connected map[enode.ID]struct{}, nodes map[enode.ID]*enode.Node) {
+	n, exist := nodes[peer] // not related event
+	if !exist {
+		return
+	}
+	_, exist = connected[peer] // already removed
+	if !exist {
+		return
+	}
+	if len(nodes) == 1 { // keep node connected if we don't have another choice
+		return
+	}
+	srv.RemovePeer(n) // remove peer permanently, otherwise p2p.Server will try to reconnect
+	delete(connected, peer)
+	if len(connected) < target { // try to connect with any other selected, but not connected node
```

Comment:
Readability suggestion:
```suggestion
	if len(connected) < target { // try to connect with any other selected (but not yet connected) node
```